오픈마켓2 [STEP 1] 아리, 제리

llghdud921 opened the pull request:
안녕하세요. 그린 @GREENOVER 
아리 @leeari95, 제리 @llghdud921 입니다!

오픈마켓2 Step1 구현을 마쳐 PR을 보냅니다.......
코드도 길어지고 기능이 많아져서 그런지 부족한 점이 많습니다 😅 

상품 상세화면은 상품 수정을 위해 최소한으로만 구현을 해둔 상태이니 이 부분 참고 부탁드립니다...!!!!
혹시 부족한 점이나 놓친 부분이 있다면 번거롭더라도 한번 더 짚어주시면 감사하겠습니다.  🙇🏻‍♀️

이번 스텝도 잘부탁드릴게요~~~ 😃

<br>

## 기능 작동 예시

새로고침|상품수정|
|:-:|:-:|
|![](https://i.imgur.com/8J8VPjy.gif)|![](https://i.imgur.com/bfbaGZi.gif)|

|이미지 추가/삭제|
|:-:|
|![](https://i.imgur.com/KPU59zV.gif)|

|키보드가 콘텐츠를 가리지 않도록 설정|
|:-:|
|![](https://i.imgur.com/j8M9vw4.gif)|

|입력검증/이미지제한|입력검증/최소이미지|
|:-:|:-:|
|![](https://i.imgur.com/hqVAYd8.gif)|![](https://i.imgur.com/KW9JYwr.gif)|

|입력검증/상품명|입력검증/상품가격|
|:-:|:-:|
|![](https://i.imgur.com/NER8GAn.gif)|![](https://i.imgur.com/xJItj0v.gif)|

|입력검증/상품내용-최소|입력검증/상품내용-최대|
|:-:|:-:|
|![](https://i.imgur.com/bDPfuyV.gif)|![](https://i.imgur.com/my8nHYv.gif)|

<br>

## 고민했던 점

### 1. 상품등록 화면 구현시 이미지를 추가하는 기능

- CollectionView의 Cell과 Header를 활용하여 View를 구성
- 이미지가 5장이 채워졌을 때 경고얼럿을 띄우도록 구현
- 사용자 입장에서 몇개의 이미지가 추가되었는지 시각적으로 확인할 수 있도록 Label 추가 구현
- 이미지 파일 용량이 300KB 이상일 경우 `UIGraphicsImageRenderer`를 이용하여 20퍼센트씩 resize

### 2. 입력 검증

- 요구사항과 API 문서를 참고하여 사용자가 제대로 입력하지 않은 부분이 있다면 얼럿을 통해 어떤 부분이 잘못되었는지 알려줄 수 있도록 구현하였습니다.

### 3. RegisterViewController을 재활용

- 화면 전환 시 상품 등록, 상품 수정 flag를 전달하여 해당 flag에 대한 분기처리를 통해서 ViewController를 재활용하였습니다.
- 이 과정에서 등록과 수정의 의미를 같이 내포할 수 있는 `EditViewController` 네이밍을 지정하였습니다.

### 4. DataSource를 ViewController와 분리

- 점점 커져가는 ViewController를 다이어트 시키기 위해 DataSource를 따로 타입으로 빼두어 ViewController에 주입시키도록 구현하였습니다.

### 5. 키보드가 콘텐츠를 가리는 현상 해결

- ScrollView와 NotificationCenter 두가지를 활용
- ScrollView의 `contentInset`을 `keyboardFrame 사이즈의 높이`만큼 버퍼를 추가하여 스크롤뷰를 확장하므로써 키보드 가림현상을 해결
    - `scrollIndicatorInsets`도 같이 변경해주도록 하였습니다.
- TextView에서 줄바꿈을 하면서 커서가 내려갈 때 따라갈 수 있도록, `ScrollEnabled` 기능을 `false`로 비활성화
- 이후 오토레이아웃으로 TextView의 높이가 늘 고정되어있는 것이 아니라 늘어날 수도 있도록 `priority`를 조절.
    - 텍스트뷰가 안에 Text가 길어질 수록 높이가 늘어나고, 그에 따라 스크롤도 자동으로 내려옵니다.

### 6. View와 Controller간의 소통

- ImageHeaderView와 RegisterViewController, 각각의 ViewController들 사이에 소통을 위해서 여러 선택지를 두고 고민하였습니다.
- 다수의 객체들에게 동시에 이벤트의 발생을 알려줄 수 있고 객체 간의 연결을 줄일 수 있는 `NotificationCenter`를 사용하였습니다.

### 7. 데이터가 변화할 때 Update하는 기능

- 상품 등록, 상품 수정 시 각 ViewController에 NotificationCenter.post하여 update 이벤트를 전달
- 이벤트 전달 시 API에 `request`하여 데이터를 업데이트
- MainViewController의 CollectionView를 아래로 쓸어내렸을 때 새 데이터를 받아오는 기능 추가 구현

<br>

## 궁금했던 것 / 조언이 필요한 점

### 1. Delegate vs Notification

`view`와 `Controller` 사이에 소통하는 방식으로 `delegate pattern`을 사용하였는데, `DataSource` 파일을 분리하면서 `DataSource`에서 `delegate` 객체를 이용하는 것에 한계를 느껴 `Notificationcenter`를 사용하였습니다. 

`NotificationCenter`의 장점도 있지만 View와 Controller 간의 소통은 delegate를 많이 활용하는 것으로 알고있는데요. 현재 코드가 적절한지 아니면 또 다른 방법을 사용할 수 있을지 궁금합니다.

### 2. ViewController에 addGestureRecognizer를 하면 발생하는 증상

- 키보드를 내리는 방법 중... `Recognizer`방법을 사용하면 CollectionView의 `Delegate` 메소드 호출이 먹통이 되는데 어떤 문제인지 전혀 파악이 안되서 어려움을 겪었습니다.. 결국엔 해결방안을 찾지 못하고 다른 방법으로 문제를 해결하였는데요. 이 부분에 대해서 조언이 필요합니다.
    
    ```swift
    // Recognizer를 활용해서 뷰컨을 터치하면 키보드 사라지는 메소드
    func hideKeyboard() {
        let tap = UITapGestureRecognizer(target: self, action: #selector(UIViewController.dismissKeyboard))
        view.addGestureRecognizer(tap)
    }
    
    @objc func dismissKeyboard() {
        view.endEditing(true)
    }
    ```
    

### 3. Index out of range

- cell을 dequeueReusable 해주는 구간에서 productList를 통해 셀을 구성하고 있는데, MainViewController을 업데이트 할 때 간헐적으로 `out of range` 에러가 나는데, 정확한 에러 원인을 찾지 못했습니다.
- 디버깅을 통해 예측해본 결과, CollectionView를 reload하는 시점이 네트워킹을 하는 시점보다 빠르기 때문에 이런 에러가 나는 것 같은데요. 이 부분을 어떻게 처리를 해줘야할지.. 좋은 방법이 떠오르지가 않네요. 조언 부탁드리겠습니다 🙇🏻‍♀️

### 4. 점점 비대해지는 ViewController를 어떻게 다이어트 시켜야하는지...

- 정말 이번 프로젝트하면서 느끼는 거지만 MVC의 단점을 몸소 느끼고 있습니다. MVVM 패턴을 익히지 않으면 해당 문제는 해결하기 어려울까요? 저희가 최대한 분리할 수 있는 부분은 분리해준 것 같은데, 혹시 부족한 부분이 보이신다면 꼭 조언 부탁드리겠습니다!

### 5. 이미지 캐싱

- 저번 스텝과 마찬가지로 Detail 부분에서도 이미지 캐싱을 진행해주고 있는데요. 캐싱을 해본 경험이 별로 없어서 이 부분은 적절한지... 궁금합니다!

### 6. UIImagePickerController

- 요구사항대로 UIImagePickerController를 사용해서 앨범에 접근할 수 있도록 구현해보았는데요. 사진을 여러장 선택하는 기능은 없는거 같고, 기본 기능만 있는 것 같더라구요. 혹시 그린은 어떤 타입을 주로 활용하셨었는지, 혹은 추천 해주실만할 타입이 있는지? 궁금합니다!

### 7. View에 들어가는 문자열 관리

- 보통 View에 삽입되는 문자열같은 경우에는 따로 네임스페이스에 빼두는 편인걸까요? 단순히 View를 구성하는 곳에 text 설정을 해주기 위한 문자열은 하드코딩 되는 것이 오히려 낫다는 생각도 드는데요. 이런 부분들도 따로 네임 스페이스로 빼서 관리를 해주어야 하는건지 그린의 생각이 궁금하네요!
    - 예시) 얼럿에 title 파라미터 “Cancel”, “취소", “수정” 등등...

<br>

하하.. 적다보니 많아졌는데.. 이번에도 잘부탁드립니다~~~😄 🍏🙏🏻

llghdud921 committed:
style: RegisterProduct파일 storyboard로 뷰 생성 #13

leeari95 committed:
feat: 컬렉션뷰에 추가할 커스텀 셀을 생성 #13

- ImageCell 타입과 XIB파일 생성
- 셀의 레이아웃을 설정하기 위해 FlowLayout 설정
- 컬렉션뷰 스크롤바를 제거

llghdud921 committed:
feat: 컬렉션뷰에 추가할 HeaderView 생성 #13

- ImageAddHeaderView 타입과 Xib 생성
- 셀의 레이아웃 설정하기 위해 FlowLayout 설정

leeari95 committed:
feat: headerView를 터치 시 ImagePickerController를 띄우도록 구현 #13

- NotificationCenter를 활용하여 headerView의 이벤트를 ViewController에 전달
- 이미지를 추가한 후 performBatchUpdates 메소드를 활용하여 collectionView 셀 추가

llghdud921 committed:
feat: notification을 이용하여 headerview내 data 변경 #13

- Notification.name extension 추가
- ImageAddHeaderView에 updateLabelText() 추가
- ImageAddHeaderView의 button에서 button과Label로 구성 변경

leeari95 committed:
feat: 이미지의 사이즈를 줄여서 데이터 크기를 압축하는 기능 추가 #15

- UIImage extension하여 resized, compress 메소드 생성

llghdud921 committed:
feat: 이미지 최대 개수 제한 기능 추가 #15

leeari95 committed:
feat: 상품등록 창에 Done 버튼을 추가 #16

- Done버튼을 눌렀을 때 이미지가 없을 경우 얼럿을 띄우도록 구현
- 기존 Alert 메세지 Message 타입부분으로 리팩토링
- add버튼 클릭시 full screen으로 뜨도록 구현

llghdud921 committed:
feat: 상품 등록 페이지 내 textField, textView 입력검증 구현 #16

leeari95 committed:
refactor: 입력검증 하는 부분을 메소드로 분리 #16

- 가독성을 높이기 위해 조건문을 따로 변수로 할당
- 메세지 부분 줄바꿈 추가 수정

llghdud921 committed:
feat: 상품 등록 화면 내 view placeholder 추가 #13

leeari95 committed:
feat: ProductCell에 setUpAccessibility 메소드 추가 #15

- 글씨가 실시간으로 커질 수 있도록 다이나믹 타입 적용

llghdud921 committed:
feat: 상품등록 네트워크하는 기능 추가 #14

- identifier 변경
- ProductRegistration 생성하는 기능 추가

leeari95 committed:
refactor: 상품등록 창에서 Done 버튼 클릭시 얼럿띄우고 창을 닫도록 개선 #15

- UIViewController+extension 수정
- request와 네트워킹 하는 부분을 메소드로 분리

leeari95 committed:
feat: 상품 등록 시 키보드 타입, 키보드가 나타날 때 스크롤 되도록 구현 #15

- RegisterProduct에 ScrollView를 배경으로 깔도록 구현
- 스택뷰의 높이의 우선순위를 낮추고 텍스트뷰의 높이 우선순위를 높임
- 텍스트뷰의 스크롤 기능을 제거
- 텍스트뷰의 높이가 늘어날 수록 스택뷰의 높이도 늘어나도록 오토레이아웃 수정
- NotificationCenter를 활용하여 키보드가 나타날 때 콘텐츠를 가리지 않도록 메서드 추가
- 스크롤을 하거나 텍스트필드 사이를 터치했을 때 키보드가 내려가도록 기능 추가

llghdud921 committed:
feat: RegisterViewController가 delegate pattern을 적용하도록 수정 #15

- ImageAddHeaderViewDelegate 생성
- 기존의 notification에서 delegate pattern으로 수정

leeari95 committed:
style: View 내부를 폴더로 분리

llghdud921 committed:
refactor: alert창에 띄우는 메세지 namespace로 구현 #15

leeari95 committed:
fix: 상품 등록시 할인가격, 재고수량을 입력안해도 기본값 0으로 들어가도록 개선 #17

llghdud921 committed:
fix: CollectionView내 cell itemisze 고정 #17

leeari95 committed:
feat: 상품등록을 완료한 후 메인화면을 업데이트 할 수 있도록 기능 구현 #15

- 기존 seyUpData 메소드를 삭제하고 requestProducts 메소드에 completion 파라미터 추가
- requestProducts를 호출하는 곳에서 다른 로직으로 구성하도록 수정
- MainViewController에 NotificationCenter를 추가
- 기존 RegisterViewController에 dismiss부분 리팩토링
- 사용하지 않는 Notification.Name 삭제

llghdud921 committed:
feat: 상세페이지 View 구현 #13

leeari95 committed:
feat: Product 스토리보드에 DetailViewController 연결 #13

- DetailViewController 생성
- IBOutlet, IBAction 연결
- MainViewController에 CollectionView delegate 메소드 추가
- performSegue를 활용하여 뷰를 전환하도록 구성

llghdud921 committed:
feat: productId를 받아 상세정보 네트워크 기능 구현 #14

- product 받아서 label text 띄우는 기능

leeari95 committed:
fix: 상품 상세 내용을 적지 않았는데, 통과되는 버그 개선 #16

- Segue가 정상적이지 않은 부분 if문으로 분기처리하여 개선

llghdud921 committed:
refactor: RegisterDataSource 파일 분리 #18

- RegisterDataSource class 생성하여 기존의 CollectionViewDataSource 내 코드 분리
- RegisterDataSource에 images 등 데이터 다루는 코드 분리
- delegate pattern 적용된 부분 notification 적용으로 개선

leeari95 committed:
feat: 상품상세정보에서 Edit > 수정 버튼을 클릭했을 때 상품등록창을 수정하도록 기능 추가 #19

- DetailViewController에서 뷰가 전환될 때 데이터를 전달하도록 prepare 메소드 추가
-  RegisterViewController에 모드 타입을 추가
- 모드에 따라 뷰의 동작 일부분을 작동하지 않도록 guard문 추가
- 모드가 modify이면 전달받은 상품정보로 뷰를 구성하도록 추가

llghdud921 committed:
feat: refresh control을 이용해서 새로고침 구현 #15

- notification deinit 코드 제거

leeari95 committed:
feat: DetailViewController에 이미지 캐싱 기능 추가 #15

- product 데이터를 전달받아서 이미지를 캐싱한 후 images에 append하도록 구현

llghdud921 committed:
feat: detailViewController에서 Images 전달받아 수정화면에 이미지 추가 #19

leeari95 committed:
feat: 상품 수정으로 이동했을 때 이미지의 버튼을 숨기도록 개선 #17

- RegisterDataSource의 프로퍼티를 private으로 접근제어 설정
- 이후 프로퍼티를 직접 수정하는 부분을 메소드로 수정할 수 있도록 추가
- 기존에 RegisterViewController에 있던 Mode 타입을 RegisterDataSource로 이동
- ImageCell에 deleteButton IBOutlet 연결
- 상품상세화면의 Segue Kind와 Presentation을 Full Screen으로 변경

llghdud921 committed:
feat: 상품수정, 상세화면에 description 보여주는 기능 추가 #19

- Product 구조체에 description 추가
- DetailViewController에 descriptionLabel 추가

leeari95 committed:
feat: 상품수정 모드에서 Done버튼을 클릭했을 때 PATCH를 보내도록 구현 #19

- doneButtonTapped에 네트워킹을 분기처리 하도록 수정
- requestModification, requestModify 메소드 추가
- 기존 createRequest 메소드명을 requestRegister로 수정
- Message에 변수 추가
- TextFieldsStackView에 createModification 메소드 추가

llghdud921 committed:
feat: 상품 수정 시 상세페이지, 메인페이지 업데이트하도록 구현 #19

- detailViewController에 notification 추가
- 상품 수정 후에 dismissUpdateDetail메서드를 통해 notification post

leeari95 committed:
feat: DetailViewController에 상품 수정을 클릭했을 때 비밀번호를 확인하는 기능 추가 #19 #15

- 얼럿을 통해 비밀번호 여부를 확인하는 코드 추가
- 컨벤션 안맞는 부분 수정처리
- UIViewController+extension에 alertInputPassword 메소드 추가
- ProductModification에 extension으로 이니셜라이저 추가

llghdud921 committed:
style: 각 페이지별 View, controller 폴더 구조를 변경

leeari95 committed:
refactor: RegisterViewController > EditViewController로 이름 수정 및 리팩토링 #18

- 폴더 및 ViewController, Storyboard 네이밍을 수정
- EditViewController 내부 접근제어 추가
- 메소드를 호출 순서로 위치를 조정하여 개선

llghdud921 committed:
refactor: MainViewController내 dataSource 코드 분리 #18

- MainViewController 호출순서로 코드 순서위치 변경

leeari95 committed:
refactor: DetailViewController 내부 리팩토링 #18

- 접근제어 추가
- 메소드 순서를 호출되는 순서, 연관성으로 개선

llghdud921 committed:
refactor: Utility를 만들어 하드코딩된 상수 namespace에 분리 #18

- Alertmesssage enum namespace 생성 및 관련 코드 수정
- AlertConstatnt enum namespace 생성 및 관련 코드 수정
- EditMode enum namespace 생성 및 관련 코드 수정
- SegueIdentifier enum namespace 생성 및 관련 코드 수정
- 상품 수정, 등록 title 관련 로직 수정

leeari95 committed:
refactor: 상품 수정 시 확인한 secret을 Done 버튼을 클릭할 때에도 활용할 수 있도록 개선 #19 #18

- alert Action Sheet를 UIViewController+extension 파일에 메소드로 분리
- alertActionSheet 메소드를 활용하여 상품수정 부분 개선
- DetailViewController에 secret 프로퍼티를 추가하고, prepare시 넘겨주도록 개선

llghdud921 committed:
docs: Update README.md

GREENOVER commented on OpenMarket/OpenMarket/Extension/Notification.Name+extension.swift:
```diff
@@ -0,0 +1,15 @@
+extension Notification.Name {
+    static let addButton = Notification.Name("addButton")
+    static let editImageCountLabel = Notification.Name("editImageCountLabel")
+    static let updataMain = Notification.Name("updateMain")
```
`update`가 맞을것 같습니다🤔

llghdud921 replied:
ㅎㅎㅎ 오타네요,,ㅎ 수정하였습니다~~

GREENOVER commented on OpenMarket/OpenMarket/Extension/UIImage+extension.swift:
```diff
@@ -0,0 +1,32 @@
+import UIKit
+
+extension UIImage {
+    func resized(percentage: CGFloat) -> UIImage {
```
혹시 이 메서드가 compress 외 다른 부분에서도 불릴 일이 있나요?🙋🏻
제한적이게 compress에서만 쓴다면 private하게 가도 될것 같아서요🤔

llghdud921 replied:
네 맞습니다! 꼼꼼하게 봐주셔서 감사드립니다 :)

GREENOVER commented on OpenMarket/OpenMarket/Extension/UIImage+extension.swift:
```diff
@@ -0,0 +1,32 @@
+import UIKit
+
+extension UIImage {
+    func resized(percentage: CGFloat) -> UIImage {
```
resized는 리사이즈 되었다로 과거형으로 해석이 되는데 로직을 보면 리사이즈를 해주는 구현 같습니다.
이에 resize가 더 적합하지 않을까 싶습니다🙋🏻

llghdud921 replied:
네이밍 수정 완료하였습니다~

GREENOVER commented on OpenMarket/OpenMarket/Extension/UIImage+extension.swift:
```diff
@@ -0,0 +1,32 @@
+        }
+        while compressedImageData.count > maxDataSize {
+            quality *= 0.8
+            compressImage = compressImage.resized(percentage: quality)
```
compress된 이미지를 뜻하는 compressedImage가 더 적합한 네이밍이 아닐까 싶어요🤔

llghdud921 replied:
네이밍 수정 완료하였습니다~ 감사합니다

GREENOVER commented on OpenMarket/OpenMarket/Extension/UIImage+extension.swift:
```diff
@@ -0,0 +1,32 @@
+            compressedImageData = compressImage.jpegData(compressionQuality: quality) ?? Data()
+        }
+        return compressImage
+    }
```
이미지 리사이징하는 일련의 로직 과정들이 유기롭고 좋네요👍

GREENOVER commented on OpenMarket/OpenMarket/Extension/UIViewController+extension.swift:
```diff
@@ -8,10 +8,48 @@
-        let okAction = UIAlertAction(title: "OK", style: .default)
-        let alert = UIAlertController(title: nil, message: message, preferredStyle: .alert)
+    func showAlert(message: String, completion: (() -> Void)? = nil) {
+        let okAction = UIAlertAction(title: "OK", style: .default) { _ in
+            completion?()
+        }
+        let alert = UIAlertController(title: "알림", message: message, preferredStyle: .alert)
```
추후 OK, 알림과 같은 title도 파라미터로 구성하면 더 확장성있게 사용될 수 있을것 같습니다🙋🏻
다만 제가말씀드린 방향은 지금 구현한 showAlert를 여러군데에서 다양하게 쓰일때를 의미하며 아래와 같이 여러 얼럿을 만든다면 제가 말씀드린 방향은 오히려 코드만 지저분해질 수도 있으니 유의해주세요👍

leeari95 replied:
휴.. 정말 의문이였던 부분이였는데 소중한 의견 고맙습니다~🙇🏻‍♀️
참고하여 코드도 다시 수정해보았습니다. 😊🙏🏻

GREENOVER commented on OpenMarket/OpenMarket/Extension/UIViewController+extension.swift:
```diff
@@ -8,10 +8,48 @@
+        self.present(alert, animated: true, completion: nil)
+    }
+    
+    func alertActionSheet(
```
alert는 무언가 경고를 위해 창을 나타내는것으로 보통 많이 쓰이는데 액션시트가 나타나는것은 그와 조금 다른 해석 같다고 느껴집니다🤔
display 혹은 show같은 prefix가 좀 더 괜찮아 보이는데 어떠신가요?

llghdud921 replied:
네! 의견 감사합니다~~~
`showActionSheet` , `showAlertPasswordInput` 로 메소드명 개선하였습니다!

GREENOVER commented on OpenMarket/OpenMarket/Extension/UIViewController+extension.swift:
```diff
@@ -8,10 +8,48 @@
+        completion: @escaping (UIAlertAction) -> Void
+    ) {
+        let modityAction = UIAlertAction(title: titles.0, style: .default, handler: completion)
+        let deleteAction = UIAlertAction(title: titles.1, style: .destructive)
```
상품 삭제 버튼은 동작하지 않는것 같은데 맞을까요?🤔
아..! 스텝2 요구사항이군요!
그때 쓰임새를 기대해봐야겠네요🙌

leeari95 replied:
안그래도 해당 부분 빼먹은 거 발견하고 바로 코드 수정 진행해보았습니다. 
아직은 쓰지 않겠지만, 추후 사용될 거기 때문에 개선해주었습니다.
캐치 감사합니다 😁

GREENOVER commented on OpenMarket/OpenMarket/Extension/UIViewController+extension.swift:
```diff
@@ -8,10 +8,48 @@
+        message: String = AlertMessage.editProduct,
+        completion: @escaping (UIAlertAction) -> Void
+    ) {
+        let modityAction = UIAlertAction(title: titles.0, style: .default, handler: completion)
```
`modify`를 의도하셨을것 같습니다🤔

llghdud921 replied:
또 오타네요..ㅎㅎ 이번 스텝에 오타가 많네요..ㅠㅠ 신경써서 작성하겠습니다!

GREENOVER commented on OpenMarket/OpenMarket/Info.plist:
```diff
@@ -2,6 +2,8 @@
 <!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
 <plist version="1.0">
 <dict>
+	<key>NSPhotoLibraryAddUsageDescription</key>
+	<string>카메라좀 쓸게</string>
```
ㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋ무섭네요 역시...🤯
여기서 NSPhotoLibraryAddUsageDescription는 사진 앨범에 접근하기 위한 설정입니다.
이에 desc이 "사진 앨범에 접근을 허용하시겠습니까?" 같은걸로 바껴야될것 같아요.
"카메라좀 쓸께"에 적합한 설정은 NSCameraUsageDescription라고 생각합니다🙋🏻

GREENOVER replied:
아 추가로 NSPhotoLibraryAddUsageDescription와 NSPhotoLibraryUsageDescriptiond의 차이가 뭘까요?

leeari95 replied:
앗 ㅋㅋㅋㅋ 저 문구를 나중에 수정해야지 하고 그대로 두었네요.. 개선해보겠습니다!!!
그리고 둘의 차이점은...
> NSPhotoLibraryUsageDescriptiond: 읽기 전용
> NSPhotoLibraryAddUsageDescription: 쓰기 전용

이 차이인걸까요~?

GREENOVER replied:
네 맞습니다 정확히 보자면 제가 알기론 위는 사진앨범 모든 기능에 대해 부여받는거고 아래는 추가 즉 쓰기 전용이라고 알고 있어요!

GREENOVER commented on OpenMarket/OpenMarket/Views/Detail/Controller/DetailViewController.swift:
```diff
@@ -0,0 +1,154 @@
+    @IBOutlet private weak var imageView: UIImageView!
+    @IBOutlet private weak var nameLabel: UILabel!
+    @IBOutlet private weak var priceLabel: UILabel!
+    @IBOutlet private weak var bagainPriceLabel: UILabel!
```
`bargain`일까요?🙋🏻

llghdud921 replied:
하하 수정했습니다!

GREENOVER commented on OpenMarket/OpenMarket/Views/Detail/Controller/DetailViewController.swift:
```diff
@@ -0,0 +1,154 @@
+    func setUpViews(product: Product) {
+        data = product
+        setUpImages()
+        DispatchQueue.main.async {
```
비동기로 동작시켜준 이유가 궁금합니다🙋🏻
그리고 image는 왜 비동기에 포함하지 않았는지도 궁금해요🙋🏻

llghdud921 replied:
requestDetail로 API 네트워킹을 마친 후 화면에 View를 표시하기 위해 비동기로 동작시켜주었습니다.
image를 포함시키지 않은 것은 step2에서 이미지를 어떻게 표시해야할지 정해지지않아서 우선 image 관련코드는 구현하지않았습니다.

GREENOVER commented on OpenMarket/OpenMarket/Views/Detail/Controller/DetailViewController.swift:
```diff
@@ -0,0 +1,154 @@
+        }
+    }
+    
+    private func setUpNotification() {
+        NotificationCenter.default.addObserver(
+            self,
+            selector: #selector(updateDetailView),
+            name: .updateDetail,
+            object: nil)
+    }
+    
+    @objc private func updateDetailView() {
+        guard let data = data else {
+            return
+        }
+        requestDetail(productId: UInt(data.id))
+    }
```
논리적으로 결함은 없으나,
view로드 시 setUpNotification > updateDetailView > requestDetail > setUpView, setUpImages로 이어지는 이 로직 흐름이 사실 이해하기는 쉽지 않네요🥲 (순서가 조금 섞여서 그런지도 모르겠습니다.)
로직의 흐름대로 순차적으로 로직의 순서를 구성해보고 한번 리팩토링을 거쳐볼 부분이 있는지 파악해보면 좋을것 같습니다!

leeari95 replied:
다시 고민해보니 순서가 뭔가 이상하긴 하네요.
> requestDetail > setUpViews > setUpImages > setUpNotification

이 순서로 이해해주시면 될 것 같습니다. 
MainViewController prepare 메소드에서 requestDetail를 호출하여 네트워킹을 시작으로 쭉 호출되는 거라고 보시면 됩니다.
그리고 DetailViewController를 상품수정 때문에 급하게 만든거라... 
미흡한 부분이 있을 수 있습니다. STEP 2에서 좀더 보완해보도록 하겠습니다. 🙇🏻‍♀️

GREENOVER commented on OpenMarket/OpenMarket/Views/Detail/Controller/DetailViewController.swift:
```diff
@@ -0,0 +1,154 @@
+                self.setUpViews(product: product)
+            case .failure(let error):
+                DispatchQueue.main.async {
+                    self.showAlert(message: error.localizedDescription)
```
사실 이런 얼럿들은 아래와 같은
![스크린샷 2022-01-20 오후 11 02 49](https://user-images.githubusercontent.com/72292617/150353393-5776cf95-754b-45f6-a847-94bafbdd06bf.png)
얼럿 메시지를 나타내줄텐데 사용자의 입장에서는 이게 왜 이런 오류가 나는지 개발자가 아니곤 알아차릴 수 없다고 생각해요.
API request, response에서 오류나는걸 조금 더 상세한 에러 처리를 해주면 좋을것 같아요🤔

leeari95 replied:
이게 저희 앱에서 실제로 일어나는 에러인걸까요?
저희가 NetworkError 타입 내부에 LocalizedError를 채택받아서 errorDescription을 구현해줌으로써
저희가 따로 구현해둔 에러가 보여질 것이라고 판단해두었기 때문에 error.localizedDescription을 호출한 부분이거든요.
따라서 NetworkError의 errorDescription부분 설명이 좀 부족한 것일까요?

GREENOVER replied:
아 정의되지않은 디스크립션이 있는것 같다 느꼈고 이를 보기위해서 강제적으로 노출시켜봤습니다!
아마 error 케이스가 하나 누락되었을수도 있다고 예상됩니다🤔

GREENOVER commented on OpenMarket/OpenMarket/Views/Edit/Controller/EditViewController.swift:
```diff
@@ -0,0 +1,271 @@
+        guard dataSource.state != .modify else {
+            return
+        }
+        dataSource.imagesRemove(at: indexPath.item)
```
사실상 상품등록 시 추가된 이미지들 우측 상단에 x표시가 아주 예쁘게 잘 붙어있어서
이 부분을 누르면 삭제 될거라고 예상했는데 이미지 즉 Item이 didSelect될때 제거가 되더라구요!
그래서 조금 사용자 경험이 떨어지는것 같아요🥲
5개까지 등록이 가능해 등록했던걸 보고 싶어 스크롤을 하다 실수로 탭하게되면 삭제되버려 다시 등록하는 불상사가 일어나면..
흠..ㅠㅠ

leeari95 replied:
오.. 그렇네요... 이 부분 의견 적극 반영해서 기존 Delegate 메소드는 제거하고, 
NotificationCenter를 활용해서 deleteButton을 터치했을 때 이미지가 삭제되도록 개선하였습니다.
그린 감사합니다 🙇🏻‍♀️

GREENOVER commented on OpenMarket/OpenMarket/Views/Edit/Controller/EditDataSource.swift:
```diff
@@ -0,0 +1,86 @@
+
+import UIKit
+
+class EditDataSource: NSObject {
```
저는 VC를 다이어트? 시키기위해 논리적 그리고 기능적으로 유사한것끼리 이렇게 별도로 분리한것 너무 좋은것 같아요👍

GREENOVER commented on OpenMarket/OpenMarket/Views/Edit/View/ImageCell.swift:
```diff
@@ -0,0 +1,21 @@
+    static let nibName = "ImageCell"
+    
+    @IBOutlet var imageView: UIImageView!
+    @IBOutlet var deleteButton: UIButton!
```
사실상 이 버튼은 UIButton 타입이지만 현재 쓰임새는 버튼이 아닌것 같습니다🤔
혹시 구현에서 제가 놓친 부분이 있을까요?

leeari95 replied:
위에 의견주신대로 개선하였더니 잘 사용하게 되었네요! ㅎㅎㅎ 감사합니다~~😁

GREENOVER commented on OpenMarket/OpenMarket/Extension/UIViewController+extension.swift:
```diff
@@ -8,10 +8,48 @@
+        let alert = UIAlertController(title: "비밀번호를 입력해주세요", message: nil, preferredStyle: .alert)
+        alert.addTextField { userTextField in
+            resultTextField = userTextField
+        }
```
사실 비밀번호는 민감한 사항이라 대부분 secret하게 처리합니다.
이에 해당 부분에서도 secret하게 나타나야할 필요가 있다고 생각해요🤔
`userTextField.isSecureTextEntry = true`와 같이 줌으로 아래처럼 다소 비밀스럽게 텍스트를 표현할 수 있습니다🙋🏻

![스크린샷 2022-01-20 오후 11 32 20](https://user-images.githubusercontent.com/72292617/150358619-9e3c3190-3fec-42bb-a9b0-f09f757da6a3.png)

또한 이 속성은 UITextInputTraits 프로토콜에 있는 연산 프로퍼티로 여기 한번 보시면 좀 재밌는게 많습니다👍

leeari95 replied:
네 이 부분도 얼럿을 수정해줄 때 개선해보았습니다.
재밌는게 많다니.. 궁금하니 한번 프로퍼티좀 살펴봐야겠네요 😁
감사합니다!!! 🤩

GREENOVER commented on OpenMarket/OpenMarket/Views/Edit/View/ImagesCollectionView.swift:
```diff
@@ -0,0 +1,39 @@
+        register(
+            headerNib,
+            forSupplementaryViewOfKind: UICollectionView.elementKindSectionHeader,
+            withReuseIdentifier: ImageAddHeaderView.identifier)
```
```swift
register(
    headerNib,
    forSupplementaryViewOfKind: UICollectionView.elementKindSectionHeader,
    withReuseIdentifier: ImageAddHeaderView.identifier
)
```
요렇게 한줄 마지막 괄호를 띄우는게 지금까지 하신 코드스타일과 통일되보입니다🙋🏻

llghdud921 replied:
네 코드컨벤션에 맞추어 수정완료했습니다!

GREENOVER commented on OpenMarket/OpenMarket/Views/Edit/Controller/EditViewController.swift:
```diff
@@ -0,0 +1,271 @@
+        }
+    }
+    
+    private func inputValidation() -> Bool {
```
input들을 검증하는 로직같은데 inputValidation은 명사의 느낌이 강합니다.
validateInput이 더 맞을것 같습니다.
다만 저는 validate는 인증된 유효한의 느낌으로 많이 쓰기에 어떤걸 검증할때는 verify라는 단어를 사용합니다.
음.. 그래서 이건 정말 참고만 해주세요!
저라면 `verifyInputData` 혹은 `verifyInputField`를 쓸것 같아요.

leeari95 replied:
verifyInputField으로 개선해보았습니다.
네이밍 참 어렵네요,,,, 감사합니다! 🙏🏻

GREENOVER commented on OpenMarket/OpenMarket/Views/Edit/Controller/EditViewController.swift:
```diff
@@ -0,0 +1,271 @@
+    }
+    
+    private func inputValidation() -> Bool {
+        let isValidImagesCount = dataSource.images.count > .zero
+        let isValidNameCount = textFieldsStackView.nameTextField.text?.count ?? .zero >= 3
+        let isValidPriceCount = (textFieldsStackView.priceTextField.text?.count ?? .zero) > .zero
+        let isMaintainMiniMumDescriptionCount = textFieldsStackView.descriptionTextView.text.count >= 10
```
is들로 시작하는 네이밍 아주 적절하고 좋네요👍

GREENOVER commented on OpenMarket/OpenMarket/Views/Edit/Controller/EditViewController.swift:
```diff
@@ -0,0 +1,271 @@
+        &&  textFieldsStackView.descriptionTextView.text != Placeholder.description
+        let isMaintainMaxiMumDescriptionCount = textFieldsStackView.descriptionTextView.text.count < 1000
+        
+        guard isValidImagesCount else {
+            showAlert(message: AlertMessage.minimumImageCount)
+            return false
+        }
```
위와 중복되는 코드가 좀 많이 존재하는데 
```swift
if !(dataSource.images.count > .zero) {
    showAlert(message: AlertMessage.minimumImageCount)
    return false
}
```
하나의 조건문으로 해주면 isValidImagesCount 상수와 현재 guard문 선언을 한번에 합칠 수 있다고 생각드는데
어떠신가요?🤔

GREENOVER commented on OpenMarket/OpenMarket/Views/Main/Controller/MainDataSource.swift:
```diff
@@ -0,0 +1,55 @@
+extension MainDataSource: UICollectionViewDataSource {
+    func collectionView(
+        _ collectionView: UICollectionView,
+        numberOfItemsInSection section: Int) -> Int {
```
여기도 아래와 같은 파라미터 코드 컨벤션으로 통일되면 더 보기 좋을것 같습니다🙋🏻

llghdud921 replied:
넵 수정했습니다~!

GREENOVER commented on OpenMarket/OpenMarket/Views/Main/Controller/MainViewController.swift:
```diff
@@ -173,9 +173,27 @@ extension MainViewController: UICollectionViewDelegate {
+    
+    func collectionView(_ collectionView: UICollectionView, didSelectItemAt indexPath: IndexPath) {
+        performSegue(withIdentifier: SegueIdentifier.productDetailView, sender: dataSource.productList[indexPath.item])
+        collectionView.deselectItem(at: indexPath, animated: true)
```
deselectItem 이 메서드 왜 필요할까요?
없으면 원하는 동작이 안되었는지 궁금해요!
또 현재 셀을 선택할때 이게 내가 지금 터치한 즉 선택한 셀인지 음영표시 혹은 셀 전체의 색이 표시되야할것 같은데
안되더라구요ㅠ
구현사항은 아니지만 사용자 입장에서 필요하지 않을까 싶어요!
컬렉션뷰에서 이거 해주려면 어떻게 하면 좋을까요~?🤔

leeari95 replied:
오.. 찾아보니 Cell에 selectedBackgroundView가 있어서 그걸 활용해서 해결해보았습니다!
deselectItem는 TableView 시절 생각하고ㅋㅋㅋ 습관적으로 넣어버렸던 거 같네요.
의견 감사합니다~🙇🏻‍♀️

GREENOVER commented on OpenMarket/OpenMarket/Views/Utility/AlertMessage.swift:
```diff
@@ -0,0 +1,23 @@
+    static let wrongPassword = "비밀번호가 맞지 않습니다. 다시 시도해주세요."
+    static let editProduct = "상품을 편집하시겠습니까?"
+    static let dataDeliveredFail = "데이터 전달에 실패하였습니다."
+    static let productError = "상품을 불러오는데 문제가 발생했습니다."
```
fetchProductError같이 더 명확한 네이밍이면 추후 얼럿 케이스가 많아졌을때 헷갈리지 않을것 같습니다🙋🏻

leeari95 replied:
오.. 그렇네요!! 네이밍 추천 감사합니다!!!🤩

GREENOVER commented on OpenMarket/OpenMarket/Views/Utility/AlertMessage.swift:
```diff
@@ -0,0 +1,23 @@
+    static let minimumImageCount = "이미지는 최소 1장 등록해야합니다."
+    static let maximumImageCount = "이미지는 최대 5개까지 첨부할 수 있어요"
+    static let minimumNameCount = "상품명이 세글자 이상이어야 합니다."
+    static let minimumPriceCount = "상품가격이 필수적으로 입력되어야 합니다."
```
필수적으로 입력되는건 minimum하고 조금 다른 맥락이라고 생각이 듭니다🤔
다른 네이밍은 없을까요?🥲

leeari95 replied:
requiredPriceCount는 어떠신가요~?

GREENOVER replied:
@leeari95 
Count라는것과 minimum은 이어진다고 보기에 count가 들어가는게 부적합하지 않나 생각이 듭니다🤔
저 에러 메시지 자체가 나타나는 원인이 어떤걸 세는게 아닌 단순 상품가격이 기입되지 않았기에 발생하는 메시지라고 보기에
requiredPrice로 충분할것 같아요!

GREENOVER commented on README.md:
```diff
@@ -37,6 +37,12 @@
     + [Trouble Shooting](#3-3-Trouble-Shooting)
     + [배운 개념](#3-4-배운-개념)
     + [PR 후 개선사항](#3-5-PR-후-개선사항)
+- [STEP 3 : 상품 등록/수정 화면 구현](#STEP-3--상품-등록/수정-화면-구현)
```
워후... 스텝3까지 아주아주 고생많으셨어요 그 와중에 리드미까지 짱짱입니다 두분👍👍👍

GREENOVER reviewed (changes requested):
@leeari95 @llghdud921 
역시 이번에도 빠르시군요!
너무 고생 많으셨습니다🤭
정확한 기한을 잘 지켜주셨네요👍
전체적으로 로직도 잘 나눠졌고 코드 논리상으로도 문제가 없기에 이번 PR은 주로 사용자 입장에서
기능에 대해 더 보완되면 좋을점들을 리뷰해봤습니다🙌
***
### 칭찬 드리고 싶은 부분
- 전체적인 View와 VC 이외에도 로직과 파일들을 잘 쪼개고 나눠주셨네요 💯
- 기능적인 부분에서도 마이너한 몇가지 제외하곤 전부 편리하게 작동하네요 💯
- 리드미 작성 역시나 최고 💯
***
### 고민하셨던 부분에 대한 저의 의견
- 고민하셨던 부분에 대한 목차들을 쭉 보고 코드로도 봤는데 전부 이견 없습니다.
너무 잘 구현해주셨고 더 첨언 드릴 말씀은 없어서 이렇게 의견아닌 의견을 남겨봅니다😅
다소 마이너한 부분에서의 궁금증이나 의견은 리뷰 코멘트를 봐주시면 될것 같습니다!
***
### 조언이 필요하신 부분에 대한 저의 의견
1. Delegate vs Notification
- 저는 이 코드에서 노티피케이션을 사용하면서 어색함은 찾지 못했어요!
딜리게이트로 구현을 하기엔 조금 다소 껄끄럽거나 역할이 분명하지 못한 애들도 있다고 보여서 노티피케이션을 활용하신점 좋은것 같습니다.
현재 MVC구조에서는 이 구현이 가장 깔끔했지 않을까 싶습니다.
2. ViewController에 addGestureRecognizer를 하면 발생하는 증상
- 아마도 요런 부분을 키워드로 한번 찾아보시고 적용해보시면 어떨까합니다.
`becomeFirstResponder`와 `resignFirstResponder`
아마 추측으로는 resonder chain이 뭔가 꼬였다고 생각들어요.
그래서 저는 구현할때 지금 구조와는 다르지만 저 키보드를 특정 뷰컨에 국한짓지 않고 아래와 같이
```swift
public func hideKeyboard() {
  UIApplication.shared.sendAction(
    #selector(UIResponder.resignFirstResponder), to: nil, from: nil, for: nil
  )
}
```
요런식으로 전역적으로 가져가 이게 그 특정한 뷰에서 나타나는게 아닌 총 AppView에서 공통적인 기능처럼 쓰이도록 해줍니다.
3. Index out of range
- 음 고민으로 말씀하신것처럼 저도 드는 생각이 네트워킹보다 뷰를 그려주는 로직이 빠르게 동작해서 그런거외엔 없을것 같다고 보여요.
동기적으로 이를 처리해보면 어떨까 생각이 듭니다.
흠.. 조금 더 고민해봐야겠습니다. 지금 코드에서는 딱 어떻게 해야할지 감이 잘 안잡히네요🥲
4. DataSource를 ViewController와 분리
- 좋은데요?!ㅎㅎ
지금 꽤나 뷰컨을 다이어트 시키려고 파일로 많이 잘 구조적으로 분리해주셨단 생각이 들어요.
이에 저도 찬성합니다.
5. 이미지 캐싱
- 네 현재 코드로는 딱히 구현이 잘못되었거나 한 부분은 없다고 보입니다..!
궁금한건 디테일로 들어가기전 상품 리스트에서 이미 이미지들을 캐싱하는것 같은데 디테일로 들어갔을때
이미지 캐싱이 놓쳤던 부분이 있다고 판단해서 그럴 시에 다시 캐싱을 해주도록 한건가요?
요 부분은 스텝2 디테일 화면 진행하시면서 더 봐보고 얘기해보면 될것 같아요🙌
6. UIImagePickerController
- 사용하신 UIImagePicker는 아시겠지만 단일 선택 밖에 안됩니다.
이는 어쩔수없는걸로 알고 있어요.
원하시는 멀티 선택을 하려면 PHPicker를 사용해야해요.
이건 iOS14 이상에서 쓸 수 있을거고 멀티 선택이 됩니다.
다만 크롭 기능이 없습니다ㅠ
저는 크롭 기능이 필요없어서 이를 쓰긴하는데요.. (저희 최소 지원도 14이상이라..)
사실 더 좋은 라이브러리들이 있기에 외부 라이브러리를 사용할 수 있다면 사용하는것도 맞는 방향인것 같습니다!
7. View에 들어가는 문자열 관리
- 저는 네임스페이스 관리 좋아해요.
나중에 수정하기도 너무 용이하고..
다만 말씀하신데로 얼럿에 국한되었을때는 굳이 안써도될것 같아요.
얼럿에 액션 타이틀을 줄때는 네임스페이스로 불러오면 오히려 직관적이지 않을때도 있기에
저는 일단 얼럿 액션 타이틀을 줄때 네임스페이스로 관리해본적은 없습니다🤔
얼럿 자체를 하나의 타입으로 생성하고 이니셜라이저를 통해 인자를 넘겨주도록 그렇게 구현하는데,
기본적으로 들어갈 타이틀 값은 디폴트로 넣어주고 변경될 사항이 있다면 이니셜 시 전달해 변경만 해주고 있습니다!
***
전체적으로 로직도 잘 나눠져 리뷰하기에 편했습니다!
이번에는 여기보다 리뷰 코멘트로 세부적인 의견들을 많이 남겨봤습니다ㅎㅎ
이번에도 수고하셨어요!!
제가 생각할때는 기능상 어색한 부분이 있기에 보완되면 좋을것 같아서 우선 request changes로 남겨보겠씁니다🙌
수정되면 재요청 부탁드립니다🙏🏻

leeari95 committed:
refactor : UIViewController+extension 내부 리팩토링 #18

- alertInputPassword 내부에 텍스트 필드 글자를 암호화하도록 설정
- alertActionSheet에 누락된 deleteHandler를 파라미터 추가하여 개선

llghdud921 committed:
fix : CollectionView를 refreshing할 때 발생되는 에러 비동기 로직으로 개선 #17

leeari95 committed:
📝 docs: README.md 내부 Trouble Shooting 내용 추가

llghdud921 committed:
refactor: extension 코드 내 네이밍 및 접근제어자 개선 #20

leeari95 committed:
refactor: AlertConstant 타입을 제거한 후 UIViewController+extension 부분 리팩토링 #20

- 상수로 따로 관리하는 것이 아니라 직접 문자열을 대입하여 관리하도록 수정

llghdud921 committed:
refactor: alert관련 네이밍 개선 및 코드 수정

leeari95 committed:
style: 앨범에 접근할 때 보여지는 description을 수정 #20

llghdud921 committed:
refactor: bargainPriceLabel 네이밍 수정 #20

leeari95 committed:
style: actionSheet의 버튼 순서를 변경 #20

- HIG를 참고하여 얼럿시트의 버튼 순서를 변경
- 앨범에 접근할 때 보여지는 description을 재수정

llghdud921 committed:
refactor: DetailViewController내 코드 순서 정리 #20

leeari95 committed:
refactor: CollectionViewDelegate를 활용하여 이미지 삭제하는 부분 리팩토링 #20

- Notification을 활용하여 UIButton을 터치했을 때 이미지가 삭제될 수 있도록 개선

llghdud921 committed:
refactor: 코드컨벤션에 맞추어 수정 #20

leeari95 committed:
style: inputValidation 네이밍을 verifyInoutField로 개선 #20

llghdud921 committed:
feat: cell 선택 시 backgroud로 선택된 셀을 표시하는 기능 구현 #13 #15 #20

leeari95 committed:
style: AlertMessage의 변수명을 명확하게 개선 #20

leeari95 commented:
그린 ~ 이번에도 디테일하고 꼼꼼한.. 피드백 최고입니다!
피드백 꼼꼼히 반영하여 개선해보고, 또 질문했었던 문제점도 해결하였습니다!!!
수정사항은 아래와 같습니다.

## 수정사항
* AlertConstant를 `제거`하고 `UIViewController+extension` 부분 전체적으로 개선
* CollectionView에서 `refreshing`할 때 `index out of range` 에러나는 부분 비동기적으로 처리하여 해결
* 의견주신 네이밍 부분 전체적으로 개선
* 앨범에 접근할 때 보여지는 description을 수정하여 개선
* HIG를 참고하여 alert, ActionSheet의 `버튼 순서를 변경`하여 개선
* DetailViewController `메소드 순서`를 개선
* 이미지 삭제시 'x'버튼을 클릭했을 때 삭제되도록 수정하여 개선
* 셀을 선택했을 때 변화가 발생하도록 `selectedBackgroundView` 설정

DM 드린 내용은 차주 평일 저녁에 한번 같이 봐주셨으면 합니다!
Recognizer 답변도 잘 이해가 가지가 않아서 함께 부탁드리겠습니다 😭🙏🏻

이 외 특이사항 없다면 merge 부탁드리겠습니다!
감사합니다~ 😊

GREENOVER commented on OpenMarket/OpenMarket/Info.plist:
```diff
@@ -3,7 +3,7 @@
 <dict>
 	<key>NSPhotoLibraryAddUsageDescription</key>
-	<string>사진 앨범에 접근을 허용하시겠습니까?</string>
+	<string>사진 앨범에 접근하도록 허용버튼을 눌러주세요.</string>
```
사실 여담인데 현업에서는 왜 접근해야하는지 대략적인 이유를 나타내줍니다.
예를들어 "상품 이미지 등록을 위해 앨범 접근을 요청합니다" 요런식으로 주고 있긴합니다!

GREENOVER reviewed (approved):
워우.. 고생하셨습니다!
역시 정말 빠르시네요ㅎㅎ
스텝2 많은 양의 코드도 척척 아주 최고네요👍
수정사항들 확인했고 이상없음에 어프로브 > 머지하겠습니다!
스텝3도 얼추 틀은 만드셔서 금방 하실것 같네요🙌
화이팅입니다!
